src/pipecat/services/cartesia.py

```diff
@@ -4,73 +4,224 @@
+        await self.connect()
+
+    async def stop(self, frame: EndFrame):
+        await super().stop(frame)
+        await self.disconnect()
+        pass
+
+    async def connect(self):
         try:
-            self._client = AsyncCartesia(api_key=self._api_key)
-            self._voice = self._client.voices.get(id=self._voice_id)
+            self._websocket = await websockets.connect(
+                f"{self._url}?api_key={self._api_key}&cartesia_version={self._cartesia_version}"
+            )
+            self._receive_task = self.get_event_loop().create_task(self._receive_task_handler())
+            self._context_appending_task = self.get_event_loop().create_task(self._context_appending_task_handler())
         except Exception as e:
             logger.exception(f"{self} initialization error: {e}")
+            self._websocket = None
 
-    async def stop(self, frame: EndFrame):
-        if self._client:
-            await self._client.close()
+    async def disconnect(self):
+        try:
+            if self._context_appending_task:
+                self._context_appending_task.cancel()
+                self._context_appending_task = None
+            if self._receive_task:
+                self._receive_task.cancel()
+                self._receive_task = None
+            if self._websocket:
+                ws = self._websocket
+                self._websocket = None
+                await ws.close()
+            self._context_id = None
+            self._context_id_start_timestamp = None
+            self._timestamped_words_buffer = []
+            self._waiting_for_ttfb = False
+            await self.stop_all_metrics()
+        except Exception as e:
+            logger.exception(f"{self} error closing websocket: {e}")
 
-    async def cancel(self, frame: CancelFrame):
-        if self._client:
-            await self._client.close()
+    async def handle_interruption(self, frame: StartInterruptionFrame, direction: FrameDirection):
+        await super().handle_interruption(frame, direction)
+        self._context_id = None
+        self._context_id_start_timestamp = None
+        self._timestamped_words_buffer = []
+        await self.stop_all_metrics()
+        await self.push_frame(LLMFullResponseEndFrame())
+
+    async def _receive_task_handler(self):
+        try:
+            async for message in self._websocket:
+                msg = json.loads(message)
+                # logger.debug(f"Received message: {msg['type']} {msg['context_id']}")
+                if not msg or msg["context_id"] != self._context_id:
+                    continue
+                if msg["type"] == "done":
+                    # unset _context_id but not the _context_id_start_timestamp because we are likely still
+                    # playing out audio and need the timestamp to set send context frames
+                    self._context_id = None
+                    self._timestamped_words_buffer.append(["LLMFullResponseEndFrame", 0])
```

Comment:
should this be a tuple instead to match the `zip` below? `("LLMFullResponseEndFrame", 0)`

Reply:
done